src/hotspot/os/windows/os_windows.cpp

```diff
@@ -5315,6 +5315,45 @@ void os::funlockfile(FILE* fp) {
   _unlock_file(fp);
 }
 
+char* os::realpath(const char* filename, char* outbuf, size_t outbuflen) {
+  return os::win32::realpath(filename, outbuf, outbuflen);
+}
+
+char* os::win32::realpath(const char* filename, char* outbuf, size_t outbuflen) {
+
+  if (filename == nullptr || outbuf == nullptr || outbuflen < 1) {
+    assert(false, "os::realpath: invalid arguments.");
+    errno = EINVAL;
+    return nullptr;
+  }
+
+  char* result = nullptr;
+  ALLOW_C_FUNCTION(::_fullpath, char* p = ::_fullpath(nullptr, filename, 0);)
+  if (p != nullptr) {
+    if (strlen(p) < outbuflen) {
+      strcpy(outbuf, p);
+      result = outbuf;
+    } else {
+      errno = ENAMETOOLONG;
+    }
+    ALLOW_C_FUNCTION(::free, ::free(p);) // *not* os::free
+  } else {
+    // there was a error
+    // In this case, use the user provided buffer but at least check whether _fullpath caused
+    // a memory overwrite.
+    if (errno == EINVAL) {
```

Comment:
There is nothing to indicate that `_fullpath` can ever set `EINVAL` it is only specified to return null on error. This code should not check errno but can just re-try with the user-supplied buffer.

Reply:
According to [the documentation](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/fullpath-wfullpath?view=msvc-170), EINVAL can be set in some circumstances.  Those circumstances are checked earlier in the function, so I have simplified this code.

Setting EINVAL (or ENAMETOOLONG) is part of the existing documentation for os::realpath(); I am loathe to change it.

Reply:
I missed the one case where it can return EINVAL (strange that is buried in the remarks section).

Given that none of the callers of realpath ever even look at errno it seems rather pointless to set it, but that cleanup should be a seperate RFE.